### PR TITLE
scripts and tools: Update ShellCheck linter

### PR DESCRIPTION
**Pull request by hebasto**
Enable some simple ShellCheck rules.

Note for reviewers: `bash` and `shellcheck` on macOS are different from ones on Ubuntu.
For local tests the latest `shellcheck` version 0.6.0 should be used (see #15166).

### CODE REVIEW

**Comment by hebasto** [comment]
ping @Empact @practicalswift @promag @koalaman

**Comment by hebasto** [comment]
Forgot about macOS quirks (#14115). Going to fix right now.

**Comment by hebasto** [comment]
> Forgot about macOS quirks (#14115). Going to fix right now.

Fixed. OP updated with a note for reviewers.

**Comment by Empact** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/16327/commits/077b62ec53bc2a892218e6f1311b3cbf0dc7daf6 test/lint/lint-shell.sh passes on macOS 10.14.5

**Comment by hebasto** [comment]
I was wrong: the 05fa54d56b8ed61925c0a7b58d505130162263df commit is misleading. There is nothing macOS specific in SC1087, SC2206, SC2207, SC2230 and SC2236 warnings.
The point is the default `shellcheck` version on bionic is:
```
$ /usr/bin/shellcheck -V | grep version:
version: 0.4.6
```

But we use version `0.6.0` (see #15166).

The misleading commit will be removed today.

**Comment by practicalswift** [comment]
Strong concept ACK:

As an active reviewer of all shell scripts added to the project I applaud this change: it will make sure more of the bugs introduced due to "language level misunderstandings" are fixed before human review takes place. That is clearly a good thing.

Case in point: This change fixes such a bug in `contrib/verify-commits/gpg.sh`.

File: autogen.sh [hunk 1]
```diff
@@ -7,10 +7,11 @@ export LC_ALL=C
 set -e
 srcdir="$(dirname $0)"
 cd "$srcdir"
+# shellcheck disable=SC2230
 if [ -z ${LIBTOOLIZE} ] && GLIBTOOLIZE="`which glibtoolize 2>/dev/null`"; then
```
Comment (dongcarl) [inline autogen.sh]: Does `command -v` work here?

File: contrib/verifybinaries/verify.sh [hunk 2]
```diff
@@ -86,18 +86,23 @@ fi
 WGETOUT=$(wget -N "$HOST1$BASEDIR$SIGNATUREFILENAME" 2>&1)
 
 #and then see if wget completed successfully
+# shellcheck disable=SC2181
 if [ $? -ne 0 ]; then
```
Comment (dongcarl) [inline contrib/verifybinaries/verify.sh]: I believe we can just do:
```sh
if ! WGETOUT=$(wget -N "$HOST1$BASEDIR$SIGNATUREFILENAME" 2>&1); then
```

See: https://github.com/koalaman/shellcheck/issues/782

**Review by dongcarl — COMMENTED** [review]
utACK e8683bf3e08c4fbc8c4159decaaa40134dac1eb5

Strongly in favor of stricter linting with shellcheck for shell scripts. They can be very hard to debug or fail silently if they're not.

File: autogen.sh (hunk 1 above)
Comment (hebasto) [reply autogen.sh]: Yes, it does. Going to fix tonight.
Also in `command -v 2>/dev/null` the ` 2>/dev/null` is redundant, right?

**Comment by practicalswift** [comment]
When @dongcarl's comments are addressed I think these changes should be ready to go. One last thing:

These tests remain disabled after this PR:

```
SC2006 # Use $(..) instead of legacy `..`.
SC2046 # Quote this to prevent word splitting.
SC2086 # Double quote to prevent globbing and word splitting.
SC2162 # read without -r will mangle backslashes.
```

I think it makes sense to keep these disabled with one exception:

@hebasto Would it be feasible to also enable `SC2006`?

Given the problems with the legacy form and the fact that all of these problems can be avoided by simply switching to `$(…)` I think it is worth having `shellcheck` guiding PR authors. That way human reviews don't have to repeat the arguments below :-)

Quoting `shellcheck`:

---

## Use `$(...)` notation instead of legacy backticked `` `...` ``.

### Problematic code

```sh
echo "You are running on `uname`"
```

### Correct code

```sh
echo "You are running on $(uname)"
```

### Rationale

Backtick command substitution `` `...` `` is legacy syntax with several issues.

1. It has a series of undefined behaviors related to quoting in POSIX.
1. It imposes a custom escaping mode with surprising results.
1. It's exceptionally hard to nest.

`$(...)` command substitution has none of these problems, and is therefore strongly encouraged.

### Exceptions

None.

### Related resources:

* [BashFaq: Why is `$(...)` preferred over `` `...` `` (backticks)?](http://mywiki.wooledge.org/BashFAQ/082)
* [StackOverflow: What is the difference between $(command) and `` `command` `` in shell programming?](https://stackoverflow.com/questions/4708549/shell-programming-whats-the-difference-between-command-and-command)

File: autogen.sh (hunk 1 above)
Comment (hebasto) [reply autogen.sh]: @dongcarl 
> Does `command -v` work here?

Fixed.

**Comment by hebasto** [comment]
@practicalswift 
> Would it be feasible to also enable SC2006?

Done.

**Comment by practicalswift** [comment]
utACK 4dd301824bf6bc646210240884db8d9c869c4874 modulo squash + @dongcarl's comment about `shellcheck disable=SC2181`

File: contrib/verifybinaries/verify.sh (hunk 2 above)
Comment (hebasto) [reply contrib/verifybinaries/verify.sh]: Fixed.

**Comment by hebasto** [comment]
@practicalswift 
> ... + @dongcarl's comment about `shellcheck disable=SC2181`

Done.

> ... modulo squash...

What commit message would be appropriate after squashing of all commits?

**Comment by practicalswift** [comment]
@hebasto Title "Enable shellcheck rules" and description could be the squashed messages? :-)

**Comment by hebasto** [comment]
@practicalswift 
> ... modulo squash...

Squashed.

**Comment by practicalswift** [comment]
Very nice!

utACK cb2f23d55f7b8ff27719561fc446fa5a9b9657c7

**Comment by hebasto** [comment]
@practicalswift the rules in commit message are ordered now.

**Comment by practicalswift** [comment]
utACK 1ac454a3844b9b8389de0f660fa9455c0efa7140

**Comment by dongcarl** [comment]
utACK 1ac454a

**Review by fanquake — APPROVED** [review]
ACK 1ac454a3844b9b8389de0f660fa9455c0efa7140

Tested `./autogen.sh`, `contrib/install_db4.sh`, `contrib/devtools/gen-manpages.sh`, `contrib/verify-commits/gpg.sh`, `contrib/verifybinaries/verify.sh bitcoin-core-0.18.0` all still work. `test/lint/lint-all.sh` seems to pass.

**Pull request merged**